### PR TITLE
RavenDB-19015 - Replay traffic watch tool for queries

### PR DESCRIPTION
**Pull request by ArieSLV**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19015/Replay-traffic-watch-tool-for-queries

### Additional description

Add traffic watch replay tool for queries into Raven.Debug.

### Type of change

- New feature

### How risky is the change?

- Low 

### Testing 

- It has been verified by manual testing

### CODE REVIEW

**Review by grisha-kotler — APPROVED** [review]

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+                
+            }
+
+            var consumers = new Task[_threads];
+            for (var i = 0; i < _threads; i++)
+            {
+                consumers[i] = new TrafficReplay(_schema, _host, _port).Execute();
+            }
+
+            while (Task.WhenAll(producers).Wait(1_000) == false)
```
Comment (karmeli87) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: create `Task.WhenAll(producers)` only once

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+
+                while (true)
+                {
+                    TrafficWatchChange item;
+                    try
+                    {
+                        item = await TrafficChannel.Reader.ReadAsync();
+                        Interlocked.Increment(ref RequestsCount);
+                    }
+                    catch (Exception)
```
Comment (karmeli87) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: let's catch a specific exception here when the channel is completed and empty

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            if (ex != null)
+            {
+                Console.WriteLine();
+                Console.WriteLine(ex);
+            }
+
+            Environment.Exit(2);
+        }
+
+        public class TrafficWatchChange
```
Comment (karmeli87) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Is that possible to take that directly from raven?

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        }
+
+        public void Stop()
+        {
+            Console.WriteLine("Stop collection traffic watch. Exiting...");
+            Dispose();
+            Environment.Exit(0);
+        }
+
+        public void Dispose()
```
Comment (karmeli87) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs:300]: do we need that?

File: tools/Raven.Debug/CommandLineApp.cs [hunk 5]
```diff
@@ -310,5 +312,83 @@ private static void ConfigureTrafficWatchCommand()
...
+
+                    var cert = certArg.HasValue() ? certArg.Value() : null;
+                    var certPass = certPassArg.HasValue() ? certPassArg.Value() : null;
+
+                    var logTrafficWatchReply = new TrafficWatchReplay(path, cert, certPass, host, port, threads );
+
+                    Console.CancelKeyPress += (sender, args) => logTrafficWatchReply.Stop();
+
+                    using (logTrafficWatchReply)
+                        logTrafficWatchReply.Start().Wait();
```
Comment (karmeli87) [inline tools/Raven.Debug/CommandLineApp.cs]: should we use `async / await` here?

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    public class TrafficWatchReplay : IDisposable
+    {
+        private readonly string[] _trafficFiles;
+        private readonly string _schema;
+        private readonly string _host;
+        private readonly int _port;
+        private readonly int _threads;
+        
+        public static HttpClient HttpClient;
+        private static readonly Channel<TrafficWatchChange> TrafficChannel = Channel.CreateBounded<TrafficWatchChange>(1024);
```
Comment (karmeli87) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Not very familiar with the rvn tool, but wonder if it is okay for those to be `static`

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 1 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Done

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 2 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Done

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 4 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs:300]: Removed with all related stuff.

File: tools/Raven.Debug/CommandLineApp.cs (hunk 5 above)
Comment (ArieSLV) [reply tools/Raven.Debug/CommandLineApp.cs]: Done

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 6 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: System.Threading.Channels.Channel provides static methods for creating channels. So, it's okay to be static.

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 3 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Done by adding [assembly: InternalsVisibleTo("Raven.Debug, PublicKey=xxx")] to ..src\Raven.Client\Properties\AssemblyInfo.Windows.cs and ..src\Raven.Client\Properties\AssemblyInfo.Linux.cs

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 6 above)
Comment (karmeli87) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Can you break the replay in the middle and start it over in the same rvn process? or can you do multiple replays in the same process?

If you can we have those issues:
1. You might have things in the queue from previous run
2. The Write might be completed for this channel, which will make it unusable.

Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs]: Channel and HttpClient are non-static now.

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs (hunk 4 above)
Comment (ArieSLV) [reply tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs:300]: Restored it for HttpClient disposing of.

File: tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+                Console.WriteLine();
+                Console.WriteLine(ex);
+            }
+
+            Environment.Exit(2);
+        }
+
+        public void Stop()
+        {
+            Console.WriteLine("Stop collection traffic watch. Exiting...");
```
Comment (grisha-kotler) [inline tools/Raven.Debug/LogTrafficWatch/TrafficWatchReplay.cs:295]: Can we wait here gracefully by using a cancellation token?

**Review by grisha-kotler — CHANGES_REQUESTED** [review]

**Comment by arekpalinski** [comment]
@grisha-kotler asked for merging this. Please address the review comment in another PR

**Pull request merged**